File: src/Autopatch.Client/Services/AutoPatchClient.cs
using System.Collections.Concurrent;
using System.Collections.ObjectModel;
using Autopatch.Client.Models;
using Autopatch.Core;
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;

namespace Autopatch.Client.Services;

/// <summary>
/// Implementation of the AutoPatch client that manages real-time data synchronization through SignalR.
/// </summary>
/// <remarks>
/// <para>
/// Every batch from the server carries a sequence number. Batches that are older than the current state are ignored; if a batch
/// is missing, or a batch cannot be applied, the client asks the server for the full data again.
/// </para>
/// <para>
/// The client reconnects automatically and subscribes to all collections again after a reconnect.
/// </para>
/// </remarks>
/// <param name="options">Configuration options for the AutoPatch client.</param>
/// <param name="serviceProvider">Service provider for dependency resolution.</param>
public class AutoPatchClient(
    IOptions<AutoPatchConfiguration> options,
    IServiceProvider serviceProvider)
    : IAutoPatchClient
{
    private static readonly TimeSpan[] DefaultReconnectDelays =
        [TimeSpan.Zero, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(10)];

    private readonly SemaphoreSlim _connectionLock = new(1, 1);

    /// <summary>
    /// Active subscriptions keyed by method name. Read on the SignalR receive thread, written by the caller.
    /// </summary>
    private readonly ConcurrentDictionary<string, Subscription> _subscriptions = new(StringComparer.Ordinal);

    private HubConnection? _connection;
    private volatile bool _stopped;
    private volatile bool _disposed;
    private int _restarting;

    /// <inheritdoc />
    public event EventHandler<bool>? OnConnectionChanged;

    /// <inheritdoc />
    public event EventHandler<AutoPatchErrorEventArgs>? OnError;

    /// <summary>
    /// Establishes a connection to the AutoPatch server using SignalR. Calling it while already connected has no effect.
    /// </summary>
    /// <param name="cancellationToken">A token to cancel the connection operation.</param>
    /// <returns>A task that represents the asynchronous connection operation.</returns>
    public async Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        await _connectionLock.WaitAsync(cancellationToken);
        try
        {
            _stopped = false;
            _connection ??= CreateConnection();
            if (_connection.State != HubConnectionState.Disconnected)
            {
                return;
            }
            await _connection.StartAsync(cancellationToken);
        }
        finally
        {
            _connectionLock.Release();
        }

        OnConnectionChanged?.Invoke(this, true);
        await ResubscribeAllAsync();
    }

    /// <summary>
    /// Disconnects from the AutoPatch server and stops reconnecting.
    /// </summary>
    /// <param name="cancellationToken">A token to cancel the disconnection operation.</param>
    /// <returns>A task that represents the asynchronous disconnection operation.</returns>
    /// <exception cref="InvalidOperationException">Thrown when not connected to the server.</exception>
    public Task DisconnectAsync(CancellationToken cancellationToken = default)
    {
        _stopped = true;
        return _connection != null
            ? _connection.StopAsync(cancellationToken)
            : throw new InvalidOperationException("Not connected.");
    }

    /// <summary>
    /// Subscribes to real-time updates for the specified type.
    /// </summary>
    /// <typeparam name="T">The type to subscribe to for updates.</typeparam>
    /// <param name="key">Optional key to identify a specific collection of this type. If null, uses the default collection.</param>
    /// <param name="authString">Optional authentication string for subscription validation.</param>
    /// <param name="cancellationToken">A token to cancel the subscription operation.</param>
    /// <returns>A task that returns true if subscription was successful, false if rejected by server validation.</returns>
    /// <exception cref="InvalidOperationException">Thrown when not connected to the server.</exception>
    public async Task<bool> SubscribeToTypeAsync<T>(string? key = null, string? authString = null, CancellationToken cancellationToken = default)
        where T : class
    {
        var connection = _connection ?? throw new InvalidOperationException("Not connected.");
        var methodName = GetMethodName<T>(key);

        var subscription = new Subscription(serviceProvider.GetRequiredService<ObservableCollection<T>>(), typeof(T))
        {
            Key = key,
            AuthString = authString,
        };
        var current = _subscriptions.GetOrAdd(methodName, subscription);
        if (!ReferenceEquals(current, subscription))
        {
            // The collection and its handler are shared, but every call is validated with its own credentials.
            if (!await connection.InvokeAsync<bool>(AutoPatchProtocol.SubscribeMethod, current.TypeName, key, authString, cancellationToken))
            {
                return false;
            }
            lock (current)
            {
                current.Subscribers++;
            }
            return true;
        }

        RegisterHandler(connection, methodName);
        bool accepted;
        try
        {
            accepted = await connection.InvokeAsync<bool>(AutoPatchProtocol.SubscribeMethod, subscription.TypeName, key, authString, cancellationToken);
        }
        catch
        {
            RemoveSubscription(connection, methodName);
            throw;
        }

        if (!accepted)
        {
            RemoveSubscription(connection, methodName);
        }
        return accepted;
    }

    /// <summary>
    /// Unsubscribes from real-time updates for the specified type.
    /// </summary>
    /// <typeparam name="T">The type to unsubscribe from.</typeparam>
    /// <param name="key">Optional key to identify a specific collection of this type. If null, uses the default collection.</param>
    /// <param name="cancellationToken">A token to cancel the unsubscription operation.</param>
    /// <returns>A task that represents the asynchronous unsubscription operation.</returns>
    /// <exception cref="InvalidOperationException">Thrown when not connected to the server.</exception>
    public Task UnsubscribeFromTypeAsync<T>(string? key = null, CancellationToken cancellationToken = default)
        where T : class
    {
        var connection = _connection ?? throw new InvalidOperationException("Not connected.");
        var methodName = GetMethodName<T>(key);

        if (!_subscriptions.TryGetValue(methodName, out var subscription))
        {
            return Task.CompletedTask;
        }

        lock (subscription)
        {
            if (--subscription.Subscribers > 0)
            {
                return Task.CompletedTask;
            }
        }

        RemoveSubscription(connection, methodName);
        return connection.InvokeAsync(AutoPatchProtocol.UnsubscribeMethod, typeof(T).Name, key, cancellationToken);
    }

    /// <summary>
    /// Gets the tracked collection for the specified type.
    /// </summary>
    /// <typeparam name="T">The type of items in the collection.</typeparam>
    /// <param name="key">Optional key to identify a specific collection of this type. If null, uses the default collection.</param>
    /// <returns>An observable collection that is synchronized with the server data.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the type is not subscribed.</exception>
    public ObservableCollection<T> GetTrackedCollection<T>(string? key = null)
        where T : class
    {
        return _subscriptions.TryGetValue(GetMethodName<T>(key), out var subscription)
            ? (ObservableCollection<T>)subscription.TrackedCollection
            : throw new InvalidOperationException($"Type {typeof(T).Name} with key '{key ?? "default"}' is not subscribed.");
    }

    /// <summary>
    /// Stops the connection and releases it.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        _stopped = true;
        if (_connection is not null)
        {
            await _connection.DisposeAsync();
        }
        GC.SuppressFinalize(this);
    }

    private static string GetMethodName<T>(string? key)
        => AutoPatchProtocol.GetMethodName(AutoPatchProtocol.GetSubscriptionKey(typeof(T).Name, key));

    private HubConnection CreateConnection()
    {
        var config = options.Value;
        var connection = new HubConnectionBuilder()
            .WithUrl(ResolveHubUrl(config.Endpoint), http =>
            {
                if (config.AccessTokenProvider is { } accessTokenProvider)
                {
                    http.AccessTokenProvider = accessTokenProvider;
                }
                config.ConfigureHttpConnection?.Invoke(http);
            })
            .WithAutomaticReconnect(new ReconnectPolicy(config.ReconnectDelays ?? DefaultReconnectDelays))
            .Build();

        connection.Closed += HandleConnectionClosed;
        connection.Reconnected += HandleReconnected;
        connection.Reconnecting += HandleReconnecting;
        return connection;
    }

    /// <summary>
    /// Uses the endpoint as hub URL; a URL without path gets the default hub path.
    /// </summary>
    private static Uri ResolveHubUrl(string endpoint)
    {
        var uri = new Uri(endpoint, UriKind.Absolute);
        return uri.AbsolutePath is "" or "/" ? new Uri(uri, AutoPatchProtocol.DefaultHubPath) : uri;
    }

    private void RegisterHandler(HubConnection connection, string methodName)
        => connection.On<string, PatchOperation[], bool, long>(methodName, (_, operations, isInitialSet, sequence)
            => HandleAutoPatchItem(methodName, operations, isInitialSet, sequence));

    private void RemoveSubscription(HubConnection connection, string methodName)
    {
        _subscriptions.TryRemove(methodName, out _);
        connection.Remove(methodName);
    }

    /// <summary>
    /// Handles incoming AutoPatch operations from the server and applies them to the tracked collection.
    /// </summary>
    /// <param name="methodName">The method name identifying the subscription.</param>
    /// <param name="changeSet">Array of JSON Patch operations to apply.</param>
    /// <param name="isInitialSet">Indicates whether this is the initial data set.</param>
    /// <param name="sequence">The sequence number of the batch.</param>
    private void HandleAutoPatchItem(string methodName, PatchOperation[] changeSet, bool isInitialSet, long sequence)
    {
        if (!_subscriptions.TryGetValue(methodName, out var subscription))
        {
            return;
        }

        void ApplyOperations()
        {
            try
            {
                lock (subscription)
                {
                    if (isInitialSet)
                    {
                        var items = CollectionPatcher.PrepareInitialSet(subscription, changeSet);
                        subscription.Items.Clear();
                        foreach (var item in items)
                        {
                            subscription.Items.Add(item);
                        }
                        subscription.Sequence = sequence;
                        subscription.IsInitialized = true;
                        subscription.EndResync();
                        return;
                    }

                    if (!subscription.IsInitialized || sequence <= subscription.Sequence)
                    {
                        return; // Not initialized yet, or already contained in the current state.
                    }

                    if (sequence != subscription.Sequence + 1)
                    {
                        RequestFullData(subscription);
                        return;
                    }

                    foreach (var step in CollectionPatcher.Prepare(subscription, changeSet))
                    {
                        step(subscription);
                    }
                    subscription.Sequence = sequence;
                }
            }
            catch (Exception ex)
            {
                RaiseError(ex, subscription);
                RequestFullData(subscription);
            }
        }

        // Use dispatcher if configured (for WPF/UI scenarios)
        if (options.Value.Dispatcher != null)
        {
            options.Value.Dispatcher(ApplyOperations);
        }
        else
        {
            ApplyOperations();
        }
    }

    /// <summary>
    /// Ignores further batches of the subscription until the server has sent the full data again.
    /// </summary>
    private void RequestFullData(Subscription subscription)
    {
        subscription.IsInitialized = false;
        if (subscription.TryBeginResync() && _connection is { State: HubConnectionState.Connected } connection)
        {
            _ = SubscribeOnServerAsync(connection, subscription);
        }
    }

    private async Task ResubscribeAllAsync()
    {
        if (_connection is not { } connection)
        {
            return;
        }

        foreach (var subscription in _subscriptions.Values)
        {
            subscription.IsInitialized = false;
            subscription.TryBeginResync();
        }
        await Task.WhenAll(_subscriptions.Values.Select(s => SubscribeOnServerAsync(connection, s)));
    }

    private async Task SubscribeOnServerAsync(HubConnection connection, Subscription subscription)
    {
        try
        {
            if (!await connection.InvokeAsync<bool>(AutoPatchProtocol.SubscribeMethod, subscription.TypeName, subscription.Key, subscription.AuthString))
            {
                RaiseError(new InvalidOperationException("The server rejected the subscription."), subscription);
            }
        }
        catch (Exception ex)
        {
            subscription.EndResync();
            RaiseError(ex, subscription);
        }
    }

    private void RaiseError(Exception exception, Subscription? subscription)
    {
        try
        {
            OnError?.Invoke(this, new AutoPatchErrorEventArgs(
                exception,
                subscription is null ? null : AutoPatchProtocol.GetSubscriptionKey(subscription.TypeName, subscription.Key)));
        }
        catch (Exception)
        {
            // An error handler must not break the receive loop.
        }
    }

    private Task HandleReconnecting(Exception? arg)
    {
        OnConnectionChanged?.Invoke(this, false);
        return Task.CompletedTask;
    }

    private Task HandleReconnected(string? arg)
    {
        OnConnectionChanged?.Invoke(this, true);
        _ = ResubscribeAllAsync();
        return Task.CompletedTask;
    }

    /// <summary>
    /// Handles the SignalR connection closed event. Unless the client was stopped, it keeps trying to connect again.
    /// </summary>
    private Task HandleConnectionClosed(Exception? arg)
    {
        OnConnectionChanged?.Invoke(this, false);
        if (!_stopped && !_disposed && Interlocked.Exchange(ref _restarting, 1) == 0)
        {
            _ = RestartAsync();
        }
        return Task.CompletedTask;
    }

    private async Task RestartAsync()
    {
        var policy = new ReconnectPolicy(options.Value.ReconnectDelays ?? DefaultReconnectDelays);
        try
        {
            for (var attempt = 0; !_stopped && !_disposed; attempt++)
            {
                await Task.Delay(policy.GetDelay(attempt));
                try
                {
                    await ConnectAsync();
                    return;
                }
                catch (Exception ex) when (!_disposed)
                {
                    RaiseError(ex, null);
                }
            }
        }
        catch (ObjectDisposedException)
        {
            // Disposed while reconnecting.
        }
        finally
        {
            Volatile.Write(ref _restarting, 0);
        }
    }

    /// <summary>
    /// Retries forever, using the configured delays and repeating the last one.
    /// </summary>
    private sealed class ReconnectPolicy(IReadOnlyList<TimeSpan> delays) : IRetryPolicy
    {
        public TimeSpan? NextRetryDelay(RetryContext retryContext) => GetDelay(retryContext.PreviousRetryCount);

        public TimeSpan GetDelay(long attempt) => delays.Count == 0 ? TimeSpan.FromSeconds(5) : delays[(int)Math.Min(attempt, delays.Count - 1)];
    }
}

File: src/Autopatch.Client.JS/src/AutoPatchClient.ts
/**
 * AutoPatch Client for JavaScript/TypeScript
 * Connects to AutoPatch SignalR hub and manages real-time data synchronization
 */

import { HubConnection, HubConnectionBuilder, HubConnectionState, IRetryPolicy, LogLevel } from '@microsoft/signalr';
import { applyPatch, Operation as JsonPatchOperation } from 'fast-json-patch';
import {
  AutoPatchConfiguration,
  Operation,
  SubscriptionResult,
  ConnectionStatus,
  AutoPatchEventHandlers,
  Trackable,
  CollectionState
} from './types';

/** Delays between reconnect attempts; the last one is repeated forever. */
const RECONNECT_DELAYS_MS = [0, 2000, 10000, 30000];

/** What is needed to subscribe again after a reconnect. */
interface SubscriptionInfo {
  typeName: string;
  key?: string;
  authString?: string;
  resyncPending: boolean;
}

/**
 * AutoPatch client that manages real-time data synchronization with SignalR.
 *
 * Every batch from the server carries a sequence number. Batches that are already contained in the current state are
 * ignored; if a batch is missing or cannot be applied, the client requests the full data again. After a reconnect all
 * collections are subscribed again.
 */
export class AutoPatchClient {
  private connection: HubConnection | null = null;
  private config: AutoPatchConfiguration;
  private eventHandlers: AutoPatchEventHandlers;
  private collections: Map<string, CollectionState<any>> = new Map();
  private subscriptions: Map<string, SubscriptionInfo> = new Map();

  constructor(config: AutoPatchConfiguration, eventHandlers: AutoPatchEventHandlers = {}) {
    this.config = {
      autoReconnect: true,
      reconnectDelay: 3000,
      ...config
    };
    this.eventHandlers = eventHandlers;
    this.setupConnection();
  }

  /**
   * Sets up the SignalR connection with proper event handlers
   */
  private setupConnection(): void {
    const builder = new HubConnectionBuilder()
      .withUrl(this.config.endpoint, {
        timeout: 60000, // 60 seconds timeout
        ...(this.config.accessTokenFactory ? { accessTokenFactory: this.config.accessTokenFactory } : {})
      })
      .configureLogging(LogLevel.Warning);

    if (this.config.autoReconnect) {
      const retryForever: IRetryPolicy = {
        nextRetryDelayInMilliseconds: (context) =>
          RECONNECT_DELAYS_MS[Math.min(context.previousRetryCount, RECONNECT_DELAYS_MS.length - 1)]
      };
      builder.withAutomaticReconnect(retryForever);
    }

    this.connection = builder.build();

    this.connection.onclose((error) => {
      this.notifyConnectionChanged(ConnectionStatus.Disconnected);
      if (error) {
        this.reportError(new Error(`Connection closed: ${error.message}`));
      }
    });

    this.connection.onreconnecting(() => {
      this.notifyConnectionChanged(ConnectionStatus.Reconnecting);
    });

    this.connection.onreconnected(() => {
      this.notifyConnectionChanged(ConnectionStatus.Connected);
      void this.resubscribeAll();
    });
  }

  /**
   * Connects to the AutoPatch server
   */
  async connect(): Promise<void> {
    if (!this.connection) {
      throw new Error('Connection not initialized');
    }

    if (this.connection.state === HubConnectionState.Connected) {
      return;
    }

    try {
      this.notifyConnectionChanged(ConnectionStatus.Connecting);
      await this.connection.start();
      this.notifyConnectionChanged(ConnectionStatus.Connected);
    } catch (error) {
      this.notifyConnectionChanged(ConnectionStatus.Disconnected);
      throw new Error(`Failed to connect: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Disconnects from the AutoPatch server
   */
  async disconnect(): Promise<void> {
    if (!this.connection) return;

    try {
      this.notifyConnectionChanged(ConnectionStatus.Disconnecting);
      await this.connection.stop();
      this.notifyConnectionChanged(ConnectionStatus.Disconnected);
    } catch (error) {
      this.reportError(new Error(`Disconnect error: ${error instanceof Error ? error.message : 'Unknown error'}`));
    }
  }

  /**
   * Subscribes to real-time updates for a specific data type
   */
  async subscribeToType<T extends Trackable>(
    typeName: string,
    key?: string,
    authString?: string
  ): Promise<SubscriptionResult> {
    if (!this.connection || this.connection.state !== HubConnectionState.Connected) {
      throw new Error('Not connected to server');
    }

    const subscriptionKey = this.getSubscriptionKey(typeName, key);
    if (this.subscriptions.has(subscriptionKey)) {
      // The collection is shared, but every call is validated with its own credentials.
      const accepted = await this.invokeSubscribe({ typeName, key, authString, resyncPending: false });
      return accepted ? { success: true } : { success: false, message: 'Subscription rejected by server' };
    }

    const methodName = this.getMethodName(subscriptionKey);
    const info: SubscriptionInfo = { typeName, key, authString, resyncPending: false };
    this.subscriptions.set(subscriptionKey, info);
    if (!this.collections.has(subscriptionKey)) {
      this.collections.set(subscriptionKey, {
        items: [],
        isInitialized: false,
        lastUpdate: new Date(),
        subscriptionKey,
        sequence: 0
      });
    }

    this.connection.on(methodName, (_: string, operations: Operation[], isInitialData: boolean, sequence?: number) => {
      this.handleDataUpdate<T>(subscriptionKey, typeName, operations, isInitialData, sequence);
    });

    try {
      const success = await this.invokeSubscribe(info);
      if (success) {
        return { success: true };
      }
      this.removeSubscription(subscriptionKey);
      return { success: false, message: 'Subscription rejected by server' };
    } catch (error) {
      this.removeSubscription(subscriptionKey);
      return {
        success: false,
        message: `Subscription failed: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Unsubscribes from real-time updates for a specific data type
   */
  async unsubscribeFromType(typeName: string, key?: string): Promise<void> {
    if (!this.connection || this.connection.state !== HubConnectionState.Connected) {
      throw new Error('Not connected to server');
    }

    const subscriptionKey = this.getSubscriptionKey(typeName, key);

    try {
      await this.connection.invoke('UnsubscribeFromType', typeName, key || null);
      this.removeSubscription(subscriptionKey);
    } catch (error) {
      this.reportError(new Error(`Unsubscribe failed: ${error instanceof Error ? error.message : 'Unknown error'}`));
    }
  }

  /**
   * Gets the current data for a subscribed type
   */
  getCollection<T extends Trackable>(typeName: string, key?: string): T[] {
    const state = this.collections.get(this.getSubscriptionKey(typeName, key));
    return state ? state.items : [];
  }

  /**
   * Checks if a collection is initialized (has received initial data)
   */
  isCollectionInitialized(typeName: string, key?: string): boolean {
    const state = this.collections.get(this.getSubscriptionKey(typeName, key));
    return state ? state.isInitialized : false;
  }

  /**
   * Gets current connection status
   */
  getConnectionStatus(): ConnectionStatus {
    if (!this.connection) return ConnectionStatus.Disconnected;

    switch (this.connection.state) {
      case HubConnectionState.Connected:
        return ConnectionStatus.Connected;
      case HubConnectionState.Connecting:
        return ConnectionStatus.Connecting;
      case HubConnectionState.Reconnecting:
        return ConnectionStatus.Reconnecting;
      case HubConnectionState.Disconnecting:
        return ConnectionStatus.Disconnecting;
      default:
        return ConnectionStatus.Disconnected;
    }
  }

  /**
   * Handles a batch from the server.
   */
  private handleDataUpdate<T extends Trackable>(
    subscriptionKey: string,
    typeName: string,
    operations: Operation[],
    isInitialData: boolean,
    sequence?: number
  ): void {
    const state = this.collections.get(subscriptionKey);
    if (!state) return;

    const hasSequence = typeof sequence === 'number';
    if (!isInitialData) {
      if (!state.isInitialized) {
        return; // Ignore until the initial set is there.
      }
      if (hasSequence && sequence! <= state.sequence) {
        return; // Already contained in the current state.
      }
      if (hasSequence && sequence! !== state.sequence + 1) {
        this.requestFullData(subscriptionKey);
        return;
      }
    }

    try {
      const cleanedOperations = (operations ?? []).map(op => {
        const cleanOp: any = { op: op.op, path: this.transformPathToCamelCase(op.path) };
        if ('value' in op && op.value !== undefined) {
          cleanOp.value = op.value;
        }
        if ('from' in op && op.from !== undefined && op.from !== null) {
          cleanOp.from = this.transformPathToCamelCase(op.from);
        }
        return cleanOp;
      });

      // Apply to a copy, so a batch that fails halfway leaves the collection untouched.
      const current = isInitialData ? [] : state.items;
      const result = applyPatch(current, cleanedOperations as readonly JsonPatchOperation[], true, false);
      const items = result.newDocument ?? current;

      state.items.length = 0;
      state.items.push(...items);
      state.lastUpdate = new Date();
      if (hasSequence) {
        state.sequence = sequence!;
      }
      if (isInitialData) {
        state.isInitialized = true;
        const info = this.subscriptions.get(subscriptionKey);
        if (info) info.resyncPending = false;
      }

      this.eventHandlers.onDataUpdated?.(typeName, operations, isInitialData);
      this.eventHandlers.onDataReceived?.(typeName, state.items, isInitialData);
    } catch (error) {
      this.reportError(new Error(`Data update failed for ${typeName}: ${error instanceof Error ? error.message : 'Unknown error'}`));
      this.requestFullData(subscriptionKey);
    }
  }

  /**
   * Ignores further batches of a collection until the server has sent the full data again.
   */
  private requestFullData(subscriptionKey: string): void {
    const state = this.collections.get(subscriptionKey);
    const info = this.subscriptions.get(subscriptionKey);
    if (!state || !info) return;

    state.isInitialized = false;
    if (info.resyncPending || this.connection?.state !== HubConnectionState.Connected) return;

    info.resyncPending = true;
    this.invokeSubscribe(info).catch((error) => {
      info.resyncPending = false;
      this.reportError(new Error(`Resync failed for ${subscriptionKey}: ${error instanceof Error ? error.message : 'Unknown error'}`));
    });
  }

  /**
   * Subscribes all collections again (used after reconnection, when the server has forgotten all groups)
   */
  private async resubscribeAll(): Promise<void> {
    const resubscriptions = Array.from(this.subscriptions.entries()).map(async ([subscriptionKey, info]) => {
      const state = this.collections.get(subscriptionKey);
      if (state) state.isInitialized = false;
      info.resyncPending = true;

      try {
        if (!(await this.invokeSubscribe(info))) {
          this.reportError(new Error(`Resubscription rejected for ${subscriptionKey}`));
        }
      } catch (error) {
        info.resyncPending = false;
        this.reportError(new Error(`Resubscription failed for ${subscriptionKey}: ${error instanceof Error ? error.message : 'Unknown error'}`));
      }
    });

    await Promise.allSettled(resubscriptions);
  }

  private invokeSubscribe(info: SubscriptionInfo): Promise<boolean> {
    return this.connection!.invoke<boolean>(
      'SubscribeToType',
      info.typeName,
      info.key || null,
      info.authString || this.config.authString || null
    );
  }

  private removeSubscription(subscriptionKey: string): void {
    this.connection?.off(this.getMethodName(subscriptionKey));
    this.subscriptions.delete(subscriptionKey);
    this.collections.delete(subscriptionKey);
  }

  private reportError(error: Error): void {
    if (this.eventHandlers.onError) {
      this.eventHandlers.onError(error);
    } else {
      console.error('[AutoPatch]', error);
    }
  }

  /**
   * Notifies event handlers of connection status changes
   */
  private notifyConnectionChanged(status: ConnectionStatus): void {
    this.eventHandlers.onConnectionChanged?.(status);
  }

  /**
   * Transforms PascalCase property paths to camelCase to match JSON serialization
   * Example: "/2/Status" -> "/2/status", "/1/CustomerName" -> "/1/customerName"
   */
  private transformPathToCamelCase(path: string): string {
    return path.replace(/\/([A-Z][a-zA-Z]*)/g, (_match, propertyName: string) => {
      const camelCase = propertyName.charAt(0).toLowerCase() + propertyName.slice(1);
      return `/${camelCase}`;
    });
  }

  /**
   * Generates subscription key for type and optional collection key
   */
  private getSubscriptionKey(typeName: string, key?: string): string {
    return key ? `${typeName}/${key}` : typeName;
  }

  private getMethodName(subscriptionKey: string): string {
    // SignalR matches client method names case-insensitively.
    return `autopatch/${subscriptionKey.toLowerCase()}`;
  }

  /**
   * Disposes the client and cleans up resources
   */
  async dispose(): Promise<void> {
    await this.disconnect();
    this.collections.clear();
    this.subscriptions.clear();
    this.connection = null;
  }
}
